File: 1286-constrained-subsequence-sum/constrained-subsequence-sum.py

'''

like the tabulation we use in the lis , dp  store at each index ind,  what is the max constrained sum wnding at nums[ind]

the inner loop of  earlier tabulation is optimised with sliding window max

'''
import heapq as h 
class Solution:
    def constrainedSubsetSum(self, nums, k):
        maxHeap = [ ]
        h.heapify(maxHeap)
        n =len(nums)
        dp = [ nums[i] for i in range(n)]
        for i in range(n):
            #this part is essentially findin the max dp[j] in the sliding window of size k from i-1 , i-2  ....i-k 
            while(maxHeap and i-maxHeap[0][1]>k):
                h.heappop(maxHeap)
            if maxHeap :
            
                dp[i] = max(dp[i], nums[i] +  -maxHeap[0][0])
            h.heappush(maxHeap,(-dp[i],i))    
        return max(dp) 


        '''

        t : n*lok
        s  n
        '''



        

File: 1286-constrained-subsequence-sum/constrained-subsequence-sum.java
class Solution {
    public int constrainedSubsetSum(int[] nums, int k) {
    int n = nums.length;

    PriorityQueue<int[]> maxHeap = new PriorityQueue<>((int[] a, int[] b)->{ return b[0]-a[0];});


    int [] dp  = new int[n];

    int ans =Integer.MIN_VALUE;

    for(int i=0 ; i<n;i++){
        int j = i-1 ; 
        dp[i] = nums[i];
        // while(j>=0){

        //     dp[i] = Math.max(dp[i], nums[i]+ dp[j]);
        //     j-=1;
        // }

        while(maxHeap.size()> 0 && i-maxHeap.peek()[1]>k){

                maxHeap.poll();


        }

        if( maxHeap.size()>0){

            dp[i] =Math.max(dp[i], nums[i]+ maxHeap.peek()[0]);
        }
        ans  = Math.max(ans, dp[i]);

        maxHeap.offer(new int[]{dp[i], i});

    }

    return ans ; 
   
    
    
    } 



        
    
}